src/GovUk.Education.ExploreEducationStatistics.Admin/Services/Public.Data/DataSetVersionMappingService.cs

```diff
@@ -0,0 +1,136 @@
+    IUserService userService,
+    PublicDataDbContext publicDataDbContext)
+    : IDataSetVersionMappingService
+{
+    public async Task<Either<ActionResult, BatchLocationMappingUpdatesResponseViewModel>> ApplyBatchMappingUpdates(
+        Guid nextDataSetVersionId,
+        BatchLocationMappingUpdatesRequest request,
+        CancellationToken cancellationToken = default)
+    {
+        return await publicDataDbContext.RequireTransaction(async () =>
+            await userService
+                .CheckIsBauUser()
+                .OnSuccess(() => CheckMappingExists(nextDataSetVersionId, cancellationToken))
+                .OnSuccess(_ => UpdateLocationOptionMappingsBatch(nextDataSetVersionId, request, cancellationToken))
+                .OnSuccess(updateSuccessesAndFailures =>
+                {
+                    // Take all the update results from this batch of updates, and check to make sure they all
+                    // succeeded.  If any have failed, we fail the entire transaction and inform the front end of the
+                    // particular update requests that failed. 
+                    var aggregatedResult = updateSuccessesAndFailures.AggregateSuccessesAndFailures();
+
+                    return aggregatedResult
+                        .OnSuccess(updates =>
+                            new BatchLocationMappingUpdatesResponseViewModel { Updates = updates })
+                        .OnFailure<ActionResult>(errors => ValidationUtils.ValidationResult(errors));
+                }));
+    }
+
+    /// <summary>
+    /// Given a batch of Location mapping update requests, this method will return a list of either success or failure
+    /// responses for each update.
+    /// </summary>
+    private async Task<List<Either<ErrorViewModel, LocationMappingUpdateResponse>>> UpdateLocationOptionMappingsBatch(
+        Guid nextDataSetVersionId,
+        BatchLocationMappingUpdatesRequest request,
+        CancellationToken cancellationToken)
+    {
+        return await request
+            .Updates
+            .ToAsyncEnumerable()
+            .SelectAwait(async (updateRequest, index) => await UpdateLocationOptionMapping(
+                nextDataSetVersionId,
+                updateRequest,
+                index,
+                cancellationToken))
+            .ToListAsync(cancellationToken);
+    }
+
+    /// <summary>
+    /// Given a Location mapping update request, this method will return either a success or failure response.
+    /// </summary>
+    private async Task<Either<ErrorViewModel, LocationMappingUpdateResponse>> UpdateLocationOptionMapping(
+        Guid nextDataSetVersionId,
+        LocationMappingUpdateRequest updateRequest,
+        int index,
+        CancellationToken cancellationToken = default)
+    {
+        var updateJsonRequest = new JsonbPathRequest<Guid>
+        {
+            TableName = nameof(PublicDataDbContext.DataSetVersionMappings),
+            IdColumnName = "TargetDataSetVersionId",
```

Comment:
could this be `IdColumnName = nameof(DataSetVersionMapping.TargetDataSetVersionId)`

Reply:
Good suggestion - done